typed-racket-doc/typed-racket/scribblings/guide/caveats.scrbl

```diff
@@ -180,3 +180,25 @@ circumvent the protections of the type system.
 
 Contract boundaries installed for typed-untyped interaction may cause
 significant slowdowns. See @secref{contract-costs} for details.
+
+
+@section{Pattern Matching and Occurrence Typing}
+
+The @racket[match] macro does not support occurrence typing (specifically,
+negative propositions) as well as @racket[cond] and @racket[if] do.
+For example, the following function safely applies @racket[add1] to integers
+and @racket[string-length] to strings but the type checker rejects it:
+
+@examples[#:label #f #:eval the-eval
+  (eval:error
+    (define (f (x : (U String Integer))) : Integer
+      (match x
+        [(? integer?) (add1 x)]
+        [_ (string-length x)])))]
+
+The reason for the type error is that Typed Racket checks the expanded form
+of the pattern-match, and @racket[match] expands to code where the call to
+@racket[string-length] occurs before the @racket[integer?] test.
```

Comment:
I think "before" is a bad word here. Typed Racket can't see the relationship between the test and the call.